### PR TITLE
config: add memory-usage-limit

### PR DESCRIPTION
**Pull request by hicqu**
Signed-off-by: qupeng <qupeng@pingcap.com>

### What problem does this PR solve?

Add configurations for memory usage limitation.

### Check List 

Tests 

- Unit test

### Release note 
* No release note.

### CODE REVIEW

File: etc/config-template.toml [hunk 1]
```diff
@@ -45,6 +45,12 @@
 ## being filled up.
 # abort-on-panic = false
 
+## Memory usage limit for the TiKV instance. Set it to 0 means the total of system memory
+## is avaiable to TiKV.
```
Comment (rleungx) [inline etc/config-template.toml:49]:
```suggestion
## is available to TiKV.
```

File: src/config.rs [hunk 2]
```diff
@@ -2686,6 +2738,13 @@ impl TiKvConfig {
         cfg.cfg_path = tmp.path().join(LAST_CONFIG_FILE).display().to_string();
         Ok((cfg, tmp))
     }
+
+    fn default_memory_usage_limit() -> ReadableSize {
+        // TODO: is it necessary to reserve some space?
+        let reserve = 0;
```
Comment (rleungx) [inline src/config.rs:2744]: What is it used for?

File: src/config.rs [hunk 3]
```diff
@@ -2460,6 +2489,29 @@ impl TiKvConfig {
...
+                    .adjust_memory_usage_limit(self.memory_usage_limit);
+                self.raftdb
+                    .adjust_memory_usage_limit(self.memory_usage_limit);
+            }
+            cmp::Ordering::Greater => {
+                let msg = format!(
+                    "memory_usage_limit {:?} is greater than total {:?}",
+                    self.memory_usage_limit, default_memory_usage_limit
+                );
+                return Err(msg.into());
```
Comment (rleungx) [inline src/config.rs:2510]: what if print a log then use the default_memory_usage_limit?

File: src/config.rs [hunk 4]
```diff
@@ -2460,6 +2489,29 @@ impl TiKvConfig {
...
         self.pessimistic_txn.validate()?;
         self.gc.validate()?;
+
+        let default_memory_usage_limit = Self::default_memory_usage_limit();
+        if self.memory_usage_limit.0 == 0 {
+            self.memory_usage_limit = default_memory_usage_limit;
+            return Ok(());
+        }
+        match self.memory_usage_limit.0.cmp(&default_memory_usage_limit.0) {
+            cmp::Ordering::Less => {
```
Comment (rleungx) [inline src/config.rs:2499]: Should we also limit the minimum memory to avoid setting the wrong value?

Comment (hicqu) [reply src/config.rs:2499]: It's a good idea but we can't determine the minimum value easily.

**Comment by rleungx** [comment]
BTW, maybe we should support dynamically changing the high water.

**Review by nolouch — APPROVED** [review]
lgtm

**Pull request merged**